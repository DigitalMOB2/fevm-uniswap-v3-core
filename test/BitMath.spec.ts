import { expect } from './shared/expect'
import { BitMathTest } from '../typechain/BitMathTest'
import { ethers } from 'hardhat'

const { BigNumber } = ethers

describe('BitMath', () => {
  let bitMath: BitMathTest
  const fixture = async () => {
    const factory = await ethers.getContractFactory('BitMathTest')
    return (await factory.deploy()) as BitMathTest
  }
  beforeEach('deploy BitMathTest', async () => {
    bitMath = await fixture()
  })

  describe('#mostSignificantBit', () => {
    it('2', async () => {
      expect(await bitMath.mostSignificantBit(2)).to.eq(1)
    })
    it('all powers of 2', async () => {
      const results = await Promise.all(
        [...Array(255)].map((_, i) => bitMath.mostSignificantBit(BigNumber.from(2).pow(i)))
      )
      expect(results).to.deep.eq([...Array(255)].map((_, i) => i))
    })
  })

  describe('#leastSignificantBit', () => {
    it('2', async () => {
      expect(await bitMath.leastSignificantBit(2)).to.eq(1)
    })
    it('all powers of 2', async () => {
      const results = await Promise.all(
        [...Array(255)].map((_, i) => bitMath.leastSignificantBit(BigNumber.from(2).pow(i)))
      )
      expect(results).to.deep.eq([...Array(255)].map((_, i) => i))
    })
  })
})
