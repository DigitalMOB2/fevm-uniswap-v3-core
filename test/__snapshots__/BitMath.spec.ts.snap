// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitMath #leastSignificantBit gas cost of max uint128 1`] = `80056`;

exports[`BitMath #leastSignificantBit gas cost of max uint256 1`] = `79928`;

exports[`BitMath #leastSignificantBit gas cost of smaller number 1`] = `81524`;

exports[`BitMath #mostSignificantBit gas cost of max uint128 1`] = `82332`;

exports[`BitMath #mostSignificantBit gas cost of max uint256 1`] = `86696`;

exports[`BitMath #mostSignificantBit gas cost of smaller number 1`] = `63188`;
